syft/generic/object.py

```diff
@@ -56,12 +56,22 @@ def describe(self, description: str) -> "AbstractObject":
         self.description = description
         return self
 
-    def tag(self, *_tags: str) -> "AbstractObject":
+    def tag(self, *tags: str) -> "AbstractObject":
         if self.tags is None:
             self.tags = set()
 
-        for new_tag in _tags:
-            self.tags.add(new_tag)
+        # Update the owner tag index
+        for tag in tags:
+            self.tags.add(tag)
+            if self.owner is not None:
+                if self.id not in self.owner._objects:
+                    self.owner.register_obj(self)
+                if tag not in self.owner._tag_to_object_ids:
```

Comment:
nit: if `self.owner._tag_to_object_ids` was a `defaultdict(set)`, we could avoid the if